Add recursive idb support

jingjingwang opened the pull request:
This PR adds support for a new syntax: a do-until-convergence loop containing a list of special MyriaL statements. The loop terminates when there will be no update to any relation. Examples are those newly added optimizer tests.

jingjingwang commented:
I expect myself adding some more checks to limit what can be put inside of the do-until-convergence loop, but would like to get the main changes being reviewed first.

For 059673c, I'm not sure if it's the best design. I would like to add a MIN aggregate that takes a list of column indices to compare two tuples in that order, however the current MIN is a binary expression. I added a new aggregate called MULTIMIN instead with less support (e.g. fakedb) than MIN, but maybe extending MIN to an nary expression is another option? Would like to hear your thought.

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -430,6 +434,114 @@ def dowhile(self, statement_list, termination_ex):
+        op = expr[0].lower()
+        args = expr[1:]
+        if op in ["bagcomp"]:
+            for _id, expr in args[0]:
```
`expr` is shadowing a function arg

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -430,6 +434,114 @@ def dowhile(self, statement_list, termination_ex):
+                elif _id in idbs:
+                    ret += [_id]
+        elif op in ["select"]:
+            for _id, expr in args[0].from_:
```
`expr` is shadowing a function arg

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -430,6 +434,114 @@ def dowhile(self, statement_list, termination_ex):
+                    continue
+                ready = True
+                leaves = self.get_idb_leaves(expr, idbs)
+                for leaf in leaves:
```
`ready = not any(idbs[leaf].scheme() is None for leaf in leaves)`

jingjingwang replied:
Cool!

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -430,6 +434,114 @@ def dowhile(self, statement_list, termination_ex):
+                if ready:
+                    iterative_inputs = self.separate_inputs(expr, idbs, False)
+                    if len(iterative_inputs) == 0:
+                        idbs[_id].children()[1] =\
```
break line on left paren

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -430,6 +434,114 @@ def dowhile(self, statement_list, termination_ex):
+                        idbs[_id].children()[1] =\
+                            raco.algebra.EmptyRelation(raco.scheme.Scheme())
+                    elif len(iterative_inputs) == 1:
+                        idbs[_id].children()[1] =\
```
break line on left paren

senderista commented on raco/myrial/parser.py:
```diff
@@ -470,6 +477,12 @@ def p_statement_assign(p):
 
     @staticmethod
+    def p_statement_idbassign(p):
+        'statement : unreserved_id COLON EQUALS LBRACKET emit_arg_list \
```
Curious why we're introducing a new token `:=` for assignment in this context. Is there some ambiguity if we used `=`? I doubt it since the parse of the `IDBASSIGN` production calls exactly the same code as an ordinary assignment statement. In any case, I find `:=` a little confusing since it's so similar to the Datalog `:-` symbol that we're expressing as `<-`.

jingjingwang replied:
The motivation is exactly the same as your concern: I wanted to show that its an IDB :) I think `=` is good and given your previous suggest of adding `idbassign_list` as a new production, parsing should have no problem. I'll change it.

jingjingwang commented on raco/algebra.py:
```diff
@@ -381,6 +381,146 @@ def shortStr(self):
 """Logical Relational Algebra"""
 
 
+class ScanIDB(ZeroaryOperator):
```
I think it's because that the plan would have cycles if `IDBController` is the "child", then RACO would apply rules endlessly.

jingjingwang commented on raco/algebra.py:
```diff
@@ -104,7 +104,7 @@ def collectParents(self, parent_map=None):
             parent_map = {}
         for c in self.children():
-            parent_map.setdefault(c, []).append(self)
+            parent_map.setdefault(id(c), []).append(self)
```
I was following the comment and impl in https://github.com/uwescience/raco/blob/master/raco/backends/myria/myria.py#L1961 (also in function `call_compile_me()` later). I think... yes it looks like a bug to me.

jingjingwang commented on raco/algebra.py:
```diff
@@ -381,6 +381,146 @@ def shortStr(self):
+            op=self.opname(), name=self.name, sch=self.scheme())
+
+
+class IDBController(NaryOperator):
```
Yes.

jingjingwang commented on raco/algebra.py:
```diff
@@ -381,6 +381,146 @@ def shortStr(self):
+                            type(expr)))
+                agg_list.append(expr)
+            else:
+                assert isinstance(expr, expression.UnnamedAttributeRef)
```
I think it's because that these attributes are indeed unnamed, since the IDBController takes two input relations with same column types but different column names. So we don't know what to call them in a MyriaL statement. Maybe there exists a better solution..?

jingjingwang commented on raco/algebra.py:
```diff
@@ -381,6 +381,146 @@ def shortStr(self):
+        if len(agg_list) == 0:
+            agg = None
+        else:
+            agg = agg_list[0]
```
Because I want to check if there are more than 1 agg and raise an error if it happens. It's different from normal groupby-agg since we can have more than 1 agg there.

jingjingwang commented on raco/algebra.py:
```diff
@@ -1797,6 +1937,28 @@ def scheme(self):
         return None
 
 
+class UntilConvergence(NaryOperator):
```
Not really empty since `DoWhile` has a check on having >=2 children (the last one is the stopping condition) while `UntilConvergence` does not. I can make them inherit a common class with that additional check if you prefer.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -413,10 +413,40 @@ def compileme(self, leftid, rightid):
+            "argIterationInput": "%s" % args[1],
+            "argEosControllerInput": "%s" % args[2],
+            "argState": self.get_agg_json(),
+            "sync": self.recursion_mode == "SYNC"
```
The initial purpose of adding `IDBController` was for async iterations, since `DoWhile` can handle sync already (although it has to store after each iteration). Then we added the sync part to `IDBController` since it's just a small change that fits into the operator design and we think it's convenient to have both in one operator. In terms of worst/best-case performance... we can always make things happen :)

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -687,6 +717,10 @@ class MyriaSplitConsumer(algebra.UnaryOperator, MyriaOperator):
 
     def __init__(self, input):
+        assert isinstance(input, MyriaSplitProducer)
+        if not hasattr(input, 'consumers'):
```
This is for the rule `RemoveSingleSplit`, which removes a `SplitProducer` that only has one `SplitConsumer` followed by a `ShuffleProducer`. So the `SplitProducer` needs to know its consumers.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -760,11 +794,15 @@ def compileme(self, inputid):
             "argChild": inputid,
             "argPf": pf
         }
+        if hasattr(self, 'buffer_type') and self.buffer_type != "None":
```
Yep, you can tell from this line that I was a Python beginner at that time :)

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -790,6 +828,53 @@ def compileme(self, inputid):
         }
 
 
+class MyriaConsumer(algebra.ZeroaryOperator, MyriaOperator):
```
This new class is again for breaking the infinite loop of optimization. The query plan being constructed has loops such as: `IDBController` -> (sending EOIs) `Consumer` -> `EOSController` -> (sending EOSs) `Consumer` -> 'IDBController`, so I would need to break the loop to stop the recursive optimization. I ended up with making these consumers as `ZeroaryOperator` to set the optimization boundary. I agree that it's not a neat refactoring but given the current design, I'm not sure if a better solution exists. Maybe we can rename it to be more clear. What do you think?

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1216,6 +1301,22 @@ def __str__(self):
+            return expr
+        (group_list, agg) = expr.get_group_agg()
+
+        for idx in range(0, 2):
```
The 3rd child is a consumer receiving EOS from `EOSController` so it's not a part of this rule (I should just use `range(2)` for sure). For shuffle vs. partial aggregation, it's not always correct to do partial aggregation before `IDBController`, since there are tricky aggs such as `CountFilter`. Another thing is it's not always more efficient to do partial aggregation before global. For example, if all the tuples are unique, then doing a paritial `DupElim` before an `IDBController` with `DupElim` would only double the cost. I think the difference here is that the aggs in `IDBController` and group-by are not the same ones.

I didn't check if eliding the shuffle is possible since I remember that redundant shuffles are removed by some other rules, but if it's not the case (or if it doesn't apply to `IDBController`) then I will add such a check here. (Is it?)

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+            return
+        if isinstance(op, algebra.UntilConvergence):
+            for idbproducer in op.children():
+                if isinstance(idbproducer.input, MyriaIDBController):
```
It should be a `MyriaSplitProducer` that sits on top of a `IDBController`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+                idbproducers[child.input.name].input.relation_key = \
+                    child.relation_key
+            else:
+                newchildren.append(child)
```
It could be any operator except `Store`.  We start from the whole MyriaL program and check the children of the current operator. If the child is a `Store` on top of a `ScanIDB`, we discard this store and tell the IDBController to store directly. Otherwise the child, no matter which type it is, is kept.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        parent_map = {}
+        op.collectParents(parent_map)
+        for idx, child in enumerate(op.children()):
+            if not isinstance(child, MyriaSplitProducer):
```
`UntilConvergence` has a list of `MyriaSplitProducer` sthat sit on top of `IDBController`s. It is actually also a child of `MyriaConsumer`, but for breaking the infinite optimization loop (see above), I made `MyriaConsumer` as a `ZeroaryOperator` with its producer as a field but as the child (`input`).

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+            if len(child.consumers) != 1:
+                continue
+            producer = parent_map[id(child.consumers[0])][0]
+            if not isinstance(producer, MyriaShuffleProducer):
```
It is, see the newly added `__str__()`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+                elif isinstance(expr, algebra.NaryOperator):
+                    expr.args[idx] = consumer
+                if isinstance(expr, algebra.Select) and \
+                        isinstance(expr.condition, expression.boolean.GTEQ):
```
It is, but the backend can't handle it yet. Since it's on top of `count` which generates integers, `GTEQ` and `GT` should be equivalent. (I need to add it to the backend though)

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        eos_controller = MyriaEOSController()
+        eos_consumers = []
+        idb_controllers = {}
+        for idb_controller in op.args:
```
`op` is `algebra.UntilConvergence` which has `args`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+    def fire(self, op):
+        if not isinstance(op, algebra.UntilConvergence):
+            return op
+        if hasattr(op, "has_been_transformed"):
```
I'll remove this flag and check `EOSController` instead.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        return op
+
+
+class RemoveEmptyFilter(rules.Rule):
```
I'll add one check in `test_galaxy_evolution()`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1926,7 +2223,10 @@ def fragments(rootOp):
         "A shortcut to call the operator's compile_me function."
         op_id = op_ids[id(op)]
-        child_op_ids = [op_ids[id(child)] for child in op.children()]
+        if isinstance(op, MyriaConsumer):
```
See above explanation for `MyriaConsumer`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        eos_consumers = []
+        idb_controllers = {}
+        for idb_controller in op.args:
+            idb_controller.children()[2] = MyriaConsumer(None, eos_controller)
```
I see, should we make it do the right thing?

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+                if op.name not in idb_producers:
+                    idb_producers[op.name] = \
+                        MyriaSplitProducer(idb_controllers[op.name])
+                consumer = MyriaConsumer(None, idb_producers[op.name])
```
Manually setting up the pair is for breaking infinite optimizations (see above discussion). The rationale is, a `UntilConvergence` consists of a list of `SplitProducer`s, each takes an `IDBController` as the child. The corresponding `SplitConsumer`s are used as the recursive inputs of IDBs (since an IDB may occur in the bodies of many statements).

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
         return algebra.UnionAll(children)
 
 
+class FillInJoinPullOrder(rules.Rule):
```
I agree it's an awkward design, however exposing it in `DO UNTIL CONVERGENCE` is ugly as well. Ideally this kind of physical level tuning should be done by a cost-based optimizer... On the other hand, it doesn't hurt if we can declare all the joins in a program to pull from some side, it might be a good optimizations too (e.g. saving one hash table when first consuming only one child). So I ended up doing this. I've also thought about not setting this arg in RACO, only doing this manually on JSON. What do you think?

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+            return op
+        # only do it for the iterative input
+        if isinstance(op.args[1], MyriaShuffleConsumer):
+            op.args[1].input.buffer_type = buffer_type
```
I didn't add `buffer_type` as one arg of `__init__` since this information is not available until `PropagateAsyncFTBuffer`, but `MyriaShuffleProducer`s are constructed before this rule. I added a set method to set the buffer type.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        return expr
+
+
+class RemoveSingleSplit(rules.Rule):
```
I added a check in `test_cc()`.

jingjingwang commented on raco/backends/myria/myria.py:
```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+        if len(eos_consumers) == 1:
+            eos_controller.input = eos_consumers[0]
+        else:
+            eos_controller.input = MyriaUnionAll(eos_consumers)
```
I don't expect `UnionAll` to have only one child in all the other places since it's constructed from `[...] + [...]`, so adding another rule seems unnecessary.

jingjingwang commented on raco/myrial/parser.py:
```diff
@@ -409,6 +409,13 @@ def p_optional_arg_list(p):
         p[0] = p[1] or []
 
     @staticmethod
+    def p_recursion_mode(p):
```
See above discussion.

jingjingwang commented:
I've addressed most comments except for a few pending discussions:

- The `Zeroary` `MyriaConsumer` for breaking infinite optimization
- `MIN` vs. `MULTIMIN` and should they use named attribute (and what if we don't know the name)
- Join pull order: global or within do-until-convergence only, where to specify

We can discuss them in person if necessary.

jingjingwang commented on raco/expression/aggregate.py:
```diff
@@ -203,6 +203,15 @@ def typeof(self, scheme, state_scheme):
         return self.input.typeof(scheme, state_scheme)
 
 
+class MULTIMIN(NaryFunction, TrivialAggregateExpression):
```
I'm happy to see `MIN` taking multiple columns but sometimes the column names are unknown (see above discussion).

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -799,7 +806,7 @@ def compileme(self, inputid):
             "argPf": pf
         }
-        if hasattr(self, 'buffer_type') and self.buffer_type != "None":
+        if self.buffer_type is not "None":
```
You just mean `None`, right?

jingjingwang replied:
This was fixed in da94a7f above.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1883,23 +1888,33 @@ def fire(self, op):
         return op
 
+    def __str__(self):
+        return ("Store[relation](IDBScan(IDBController)) ->"
```
should be trailing space after arrow

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1883,23 +1888,33 @@ def fire(self, op):
+                "IDBController[relation]")
+
+
+def replace_child_with(op, child, replacement):
```
Why don't we `assert` in an `else` clause rather than doing nothing for unexpected operator types.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1937,51 +1952,50 @@ def fire(self, op):
+                            agg.threshold = op.condition.right.value
+                        else:
+                            # count generates integers, translate it to GTEQ
+                            agg.threshold = op.condition.right.value + 1
```
shouldn't you assert `isinstance(op.condition, expression.boolean.GT)`?

jingjingwang replied:
The if condition above ensures that it's either a `GTEQ` or `GT` so I put `else` here.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1937,51 +1952,50 @@ def fire(self, op):
-        if hasattr(op, "has_been_transformed"):
-            if op.has_been_transformed:
-                return op
+        if any(isinstance(ch, MyriaEOSController) for ch in op.args):
```
nitpick: `children()` not `args`

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -2003,14 +2017,20 @@ def fire(self, op):
+def idb_until_convergence(**kwargs):
+    ret = [DoUntilConvergence(), RemoveEmptyFilter(), StoreFromIDB(),
+           RemoveSingleSplit()]
+    if kwargs.get('async_ft', 'None') is not None:
```
Why the second argument? The default of one-arg `get()` is `None`, of course. You could really just say `kwargs.get('async_ft')`, since apparently any non-null value is equivalent.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -2043,23 +2063,20 @@ def opt_rules(self, **kwargs):
-        compile_grps_sequence.append([RemoveSingleSplit()])
 
-        if kwargs.get('join_pull_order'):
+        if kwargs.get('join_pull_order', 'None') is not None:
```
Again, the default of `get()` is `None`, so not sure what you're trying to do with the literal string `'None'`.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -2302,4 +2319,4 @@ def compile_to_json(raw_query, logical_plan, physical_plan,
     return {"rawQuery": raw_query, "logicalRa": str(logical_plan),
             "language": language, "plan": compile_plan(physical_plan),
-            "ftMode": kwargs.get('async_ft')}
+            "ftMode": kwargs.get('async_ft', None)}
```
The default of one-arg `get()` is already `None`.

jingjingwang replied:
Also fixed in da94a7f.

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1914,14 +1902,44 @@ def fire(self, op):
         self._disabled = False
-        self.pull_order = pull_order
+
+    def get_idb_side(self, op):
```
This is a misleading name: it doesn't distinguish between "sides" at all, but just tells you if there's an `IDBController` in the current operator's subtree. I think `has_idb_child()` might be a more accurate name.

jingjingwang replied:
Turned out that `preorder` doesn't work if `stop_recursion` is added to it, since finding an `MyriaIDBController` requires to pass a consumer with `stop_recursion=True`.

jingjingwang replied:
Exactly, I can't remember why I came up with such a weird name...

senderista commented on raco/backends/myria/myria.py:
```diff
@@ -1914,14 +1902,44 @@ def fire(self, op):
+            if isinstance(ch, MyriaSymmetricHashJoin):
+                left_idb = self.get_idb_side(ch.left)
+                right_idb = self.get_idb_side(ch.right)
+                if op.pull_order == 'PULL_IDB':
```
As I mentioned earlier, I think this code would be more readable if `op.pull_order` became `op.pull_order_policy`.

senderista commented on raco/myrial/emitarg.py:
```diff
@@ -96,7 +96,7 @@ def get_statemods(self):
 
     def __repr__(self):
-        return 'NaryEmitArg(%r)' % self.sexprs
+        return 'NaryEmitArg(%r,%r)' % (self.column_names, self.sexprs)
```
This is less broken than the original, but still broken. You need to include all constructor args:
```
return 'NaryEmitArg(%r,%r,%r)' % (self.column_names, self.sexprs, self.statemods)
```

senderista commented on raco/myrial/interpreter.py:
```diff
@@ -488,18 +488,16 @@ def separate_inputs(self, expr, idbs, is_init):
         return []
 
-    def untilconvergence(self, statement_list, recursion_mode):
+    def untilconvergence(self, statement_list, recursion_mode, pull_order):
```
Per above, `pull_order_policy`.

senderista commented on raco/myrial/optimizer_tests.py:
```diff
@@ -1263,22 +1264,27 @@ def test_cc(self):
-            CC = [$0, MIN($1)] <- [from V emit V.$0 as x, V.$0 as y] +
-                    [from E, CC where E.$0 = CC.$0 emit E.$1, CC.$1];
-        until convergence;
+            CC = [nid, MIN(cid) as cid] <-
```
The named attributes are way more readable!

jingjingwang replied:
Yay!

senderista commented on raco/myrial/optimizer_tests.py:
```diff
@@ -1325,18 +1331,18 @@ def test_galaxy_evolution(self):
-           emit Galaxies.$0+1, Edges.$2];
-        until convergence async;
+           where Galaxies.time = Edges.time and
+           Galaxies.gid = Edges.gid1 and Edges.num >= 4
```
It would be nice to add a test somewhere for the `GT` threshold since we support it as well.

senderista commented on raco/myrial/scanner.py:
```diff
@@ -8,7 +8,8 @@
             'OPP', 'TIPSY', 'UDA', 'TRUE', 'FALSE', 'HASH', 'BROADCAST',
-            'ROUND_ROBIN', 'UNTIL', 'CONVERGENCE', "SYNC", "ASYNC"]
+            'ROUND_ROBIN', 'UNTIL', 'CONVERGENCE', "SYNC", "ASYNC",
+            'ALTER', 'PULL_IDB', 'PULL_EDB', 'BUILD_EDB']
```
I'm slightly concerned about the `ALTER` keyword causing misunderstanding (that it means "alter" and not "alternate"). Maybe we should ask some people unfamiliar with this syntax what it suggests to them. An unambiguous *cough* alternative might be `ALTERN` or just `ALTERNATE`.

BrandonHaynes replied:
+1 `ALTERNATE`

senderista commented on raco/myrial/optimizer_tests.py:
```diff
@@ -1325,18 +1331,18 @@ def test_galaxy_evolution(self):
+        until convergence async build_EDB;
         store(Galaxies, Galaxies);
         """
         lp = self.get_logical_plan(query, async_ft='REJOIN')
```
I can't comment on the definitions of `get_logical_plan()`, `logical_to_physical()`, `compile_to_json()` here, so: don't we need to add a checkbox to MyriaWeb exposing the `async_ft` option? I can create an issue in the `myria-web` repo if so.

jingjingwang replied:
Yes we need it, and I have a `myria-web` branch. Will link to it later.

jingjingwang commented on raco/algebra.py:
```diff
@@ -66,6 +70,8 @@ def scheme(self):
     def walk(self):
         """Return an iterator over the tree of operators."""
         yield self
+        if self.stop_recursion:
```
Yep they need it as well. One concern about adding this `stop_recursion` is that I'm not sure if I have added it to every necessary place. Right now I'm only relying on tests.

jingjingwang commented:
Goes with https://github.com/uwescience/myria/pull/877.
